---
title: "assignment"
author: "Jesslyn Jesslyn"
date: "2025-04-06"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
  library(memes)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
})
```

# Task

In the same dataset of ATAC on chr19, plot 1) the insertion (i.e. ‘cuts’) profile of nucleosome-free fragments and 2) the centers of nucleosome-containing fragments, around the high-confidence motifs of **two** factors.

You can choose your own factors of interest, or for instance use KLF4 and MAZ.

Expected form of the answer: 2 figures (one for each factor), each containing the two signals around the motifs

# Download chr19 atac data

```{r, eval=FALSE}
options(timeout=1800)
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam", "atac.chr19.bam", mode = "wb")
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam.bai", "atac.chr19.bam.bai", mode = "wb")
```

# Download chr19 genome

```{r}
ah <- AnnotationHub(localHub=TRUE)
genome <- ah[["AH68356"]]
# get the sequence for chr19:
chr19 <- import(genome)["19"]
```

# Prepare tracks

```{r}
bam <- "atac.chr19.bam"

# insertion track
bam2bw(bam, output_bw = "NF_cuts.bw", paired=TRUE, binWidth=1L, type="ends", minFragLength=30, 
       maxFragLength=120, trim=4L, extend=3L, forceSeqlevelsStyle = "Ensembl")

# nucleosome-containing fragments track
bam2bw(bam, output_bw = "mono_centers.bw", paired=TRUE, binWidth=5L, minFragLength=140, trim=4L,
       maxFragLength=220, type="center", extend=10L, forceSeqlevelsStyle = "Ensembl")

# list of tracks
tracks <- list.files(pattern="bw$")
names(tracks) <- gsub("\\.bw","",basename(tracks))
```

# KLF4 motif

```{r}
# get KLF4 motif
motif_KLF4 <- MotifDb::query(MotifDb, c("KLF4","Mus"))
names(motif_KLF4)
motif_KLF4 <- motif_KLF4[[1]]
motif2_KLF4 <- convert_motifs(motif_KLF4, class="TFBSTools-PFMatrix")

# find motif matches across chr19
moi_KLF4 <- motifmatchr::matchMotifs(motif2_KLF4, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_KLF4 <- as(setNames(moi_KLF4,names(chr19)), "GRanges")
```

```{r, fig.width=4, fig.height=4}
# extract signals around the motif occurences
sm_KLF4 <- signal2Matrix(tracks, moi_KLF4, w=5, extend=300)

# plot the signals:
plotEnrichedHeatmaps(sm_KLF4, trim=0.95, multiScale = TRUE, axis_name = "0")
```

# MAZ motif

```{r}
# get KLF4 motif
motif_MAZ <- MotifDb::query(MotifDb, c("MAZ","Mus"))
names(motif_MAZ)
motif_MAZ <- motif_MAZ[[1]]
motif2_MAZ <- convert_motifs(motif_MAZ, class="TFBSTools-PFMatrix")

# find motif matches across chr19
moi_MAZ <- motifmatchr::matchMotifs(motif2_MAZ, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_MAZ <- as(setNames(moi_MAZ,names(chr19)), "GRanges")
```

```{r, fig.width=4, fig.height=4}
# extract signals around the motif occurences
sm_MAZ <- signal2Matrix(tracks, moi_MAZ, w=5, extend=300)

# plot the signals:
plotEnrichedHeatmaps(sm_MAZ, trim=0.95, multiScale = TRUE, axis_name = "0")
```